system/datatypes/mailinglist.php

```diff
@@ -412,246 +439,344 @@ private function unsubscribeAction() {
-			return ['redirect', $this->constructFullPath($this->pagename . '/' . $mailingId)];
+			return null;
 		}
 
-		// check if given id was correct
+		$mailingId = $request->getArg(1);
+
+		// check given id
 		/** @var HyphaDomElement $mailing */
 		$mailing = $this->xml->document()->getElementById($mailingId);
 		if (!$mailing instanceof HyphaDomElement) {
 			notify('error', __('not-found')); // 404
+			return null;
+		}
 
+		if (self::MAILING_STATUS_DRAFT != $mailing->getAttribute('status')) {
+			notify('error', __('unable-to-edit'));
 			return null;
 		}
 
-		// get status
-		$status = $mailing->getAttribute('status');
-		$inDraft = self::MAILING_STATUS_DRAFT == $status;
+		// create form
+		$formData = [
+			self::FIELD_NAME_ID => $mailingId,
+			self::FIELD_NAME_SUBJECT => $mailing->getAttribute('subject'),
+			self::FIELD_NAME_MESSAGE => $mailing->getHtml(),
+		];
+		$form = $this->createMailingForm($formData);
 
-		// throw error if status is draft without client logged in
-		if ($inDraft && !isUser()) {
-			notify('error', __('not-found')); // 404
+		$cancelPath = str_replace('{id}', $mailingId, self::PATH_MAILS_VIEW_ID);
+		$submitPath = str_replace('{id}', $mailingId, self::PATH_MAILS_EDIT_ID);
 
+		return $this->mailingFormViewRender($form, $cancelPath, $submitPath);
+	}
+
+	/**
+	 * no user
+	 * check mailing id
+	 * check mailing
+	 * check draft status
+	 *
+	 * @param HyphaRequest $request
+	 *
+	 * @return array|null
+	 */
+	private function mailingEditAction(HyphaRequest $request) {
+		// throw error if edit is requested without client logged in
+		if (!isUser()) {
+			notify('error', __('login-to-edit'));
 			return null;
 		}
 
-		// display mailing
-		$date = $mailing->getAttribute('date');
-		$main = $this->findBySelector('#main');
-		$main->append('<div><h2>' . $mailing->getAttribute('subject') . '</h2></div>');
-		if ($date) {
-			$date = new \DateTime($date);
-			$main->append('<div>' . __('date') . ': ' . $date->format(__('ml-date-format')) . '</h2></div>');
+		$mailingId = $request->getArg(1);
+
+		/** @var HyphaDomElement $mailing */
+		$mailing = $this->xml->document()->getElementById($mailingId);
+		if (!$mailing instanceof HyphaDomElement) {
+			notify('error', __('not-found')); // 404
+			return null;
 		}
-		if (isUser()) {
-			$receivers = $mailing->getAttribute('receivers');
-			if (null != $receivers) {
-				$main->append('<div>' . __('ml-received-by') . ': ' . $receivers . '</div>');
-			}
+
+		$postedId = $request->getPostValue(self::FIELD_NAME_ID);
+		if ($postedId != $mailingId) {
+			notify('error', __('ml-inconsistent-parameters'));
+			return null;
 		}
-		$main->append($mailing->getHtml());
 
-		// buttons
-		$commands = $this->findBySelector('#pageCommands');
-		$commands->append($this->makeActionButton(__('back'), ''));
+		// check if status is correct
+		$status = $mailing->getAttribute('status');
+		if (self::MAILING_STATUS_SENT == $status) {
+			notify('success', __('ml-successfully-sent'));
```

Comment:
Huh? When status is SENT, this throes away the edit and says "successfully sent"? What's the idea here?

Reply:
removed it